subprojects/soak/src/integTest/groovy/org/gradle/vfs/VirtualFileSystemRetentionSoakTest.groovy

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.vfs
+
+import org.gradle.integtests.fixtures.daemon.DaemonIntegrationSpec
+import org.gradle.internal.os.OperatingSystem
+import org.gradle.internal.service.scopes.VirtualFileSystemServices
+import org.gradle.soak.categories.SoakTest
+import org.gradle.test.fixtures.file.TestFile
+import org.junit.experimental.categories.Category
+
+@Category(SoakTest)
+class VirtualFileSystemRetentionSoakTest extends DaemonIntegrationSpec {
+
+    private static final int NUMBER_OF_SUBPROJECTS = 50
+    private static final int NUMBER_OF_SOURCES_PER_SUBPROJECT = 100
+    private static final int TOTAL_NUMBER_OF_SOURCES = NUMBER_OF_SUBPROJECTS * NUMBER_OF_SOURCES_PER_SUBPROJECT
+    private static final double LOST_EVENTS_RATIO_MAC_OS = 0.75
+    private static final double LOST_EVENTS_RATIO_WINDOWS = 0.1
+
+    List<TestFile> sourceFiles
+
+    def setup() {
+        // Use 20 minutes idle timeout since the test may be running longer with an idle daemon
+        executer.withDaemonIdleTimeoutSecs(1200)
+        def subprojects = (1..NUMBER_OF_SUBPROJECTS).collect { "project$it" }
+        def rootProject = multiProjectBuild("javaProject", subprojects) {
+            buildFile << """
+                allprojects {
+                    apply plugin: 'java-library'
+
+                    tasks.withType(JavaCompile).configureEach {
+                        options.fork = true
+                    }
+                }
+            """
+        }
+        sourceFiles = subprojects.collectMany { projectDir ->
+            (1..NUMBER_OF_SOURCES_PER_SUBPROJECT).collect {
+                def sourceFile = rootProject.file("${projectDir}/src/main/java/my/domain/Dummy${it}.java")
+                modifySourceFile(sourceFile, 0)
+                return sourceFile
+            }
+        }
+
+        executer.beforeExecute {
+            withArgument("-D${VirtualFileSystemServices.VFS_RETENTION_ENABLED_PROPERTY}=true")
+        }
+    }
+
+    def "file watching works with multiple builds on the same daemon"() {
+        when:
+        succeeds("assemble")
+        def daemon = daemons.daemon
+        then:
+        daemon.assertIdle()
+
+        expect:
+        (1..100).each { iteration ->
```

Comment:
Nit:
```suggestion
        100.times { iteration ->
```

Reply:
Also, we probably should extract the magic number to a constant.

Reply:
Both done.